### ISSUE TITLE
Confusing error: member reference base type '_Atomic(struct x)' is not a structure or union

### ISSUE THREAD

**Issue by pmor13** [llvm/llvm-project#54563]
```
_Atomic struct x { int x; } x;
void f(int y)
{
    x.x = y;
}
```
```
$ clang t0.c -std=c11 -pedantic -Wall -Wextra -c
<source>:4:6: error: member reference base type '_Atomic(struct x)' is not a structure or union
    x.x = y;
```
The atomic structure is still a structure.

Expected to be:
> accessing a member 'x' of an atomic structure 'x'

Or similar.

**Comment by AaronBallman** [comment]
Issuing an error is fairly reasonable behavior (accessing a member of an atomic struct or union is undefined behavior per C11 6.2.5.3p5), but the diagnostic wording itself is pretty unhelpful.

**Comment by AaronBallman** [comment]
> Is this actually UB even in an unevaluated context?

C2x 6.5.2.3p5: Accessing a member of an atomic structure or union object results in undefined behavior.

That's not clear on whether "accessing" means runtime or not, but I think it's reasonable to treat this as not being undefined behavior in an unevaluated context. Certainly the compile time lookup doesn't cause UB. WDYT?

**Comment by AaronBallman** [comment]
Hmm, I convinced myself that's worthwhile: https://reviews.llvm.org/D122681

**Comment by jyknight** [comment]
I think there's 3 situations in question to consider:
1. Actual memory operation: `int x = s.x;`
2. Address calculation (emitting no actual memory ops): `int *x = &s.x;`
3. Unevaluated access: `int size = sizeof(s.x);`

Case 1 is clearly UB -- you cannot validly mix atomic access sizes and still get correct behavior. I think case 2 probably should be considered UB as well -- while the pointer can validly be computed, someone probably is expecting to be able to dereference the result, which would be incorrect. So unless the spec _required_ this to compile, we probably don't need to allow it. I'll also note that Clang and GCC disagree on the type of the expression here -- GCC considers the result to be `_Atomic int *` (with warning), while Clang says `int*` (with warning-error).

However, case 3 seems like it should be accepted.

Also, a little quirk I noticed -- after your patch, case 2 _does_ emit an actual (unused) atomic load. That seems unexpected? E.g. compile the following with `clang -Wno-error=atomic-access`.

```
_Atomic struct S { int val; } x;
void test(void) {
  int *f = &x.val;
}
```

**Comment by AaronBallman** [comment]
> However, case 3 seems like it should be accepted.

I've landed cf4a67bcbbb109347b1e2d6cedf69bae142a182f to address the third point.

> Also, a little quirk I noticed -- after your patch, case 2 does emit an actual (unused) atomic load. That seems unexpected? E.g. compile the following with clang -Wno-error=atomic-access.

That's because we add the implicit cast from atomic to nonatomic here: https://github.com/llvm/llvm-project/commit/3c84e4a0dbd08fc03bbcdd8354a984e0efcf7672#diff-666d45282070e66014f8fb70e6c089762ea67ad4891f9819fdf5f8a5d5a0f570R1303

However, I believe we need to correct the base expression in order for the rest of the semantic analysis to continue properly (without the cast, I get crashes later).

**Comment by jyknight** [comment]
(Reviving this old thread because I just found this tab sitting open in my browser after a few years...sorry. ;p)

So, in `&x.val`, CK_AtomicToNonAtomic cast is not just changing the type, but actually copying the data into a temporary, and then referring to the address of the temporary. If we have this as a disablable warning (do we even need that, vs hard error?), that seems not good.

Maybe this should instead be creating an ImplicitCastExpr with CK_NoOp cast type? That appears to have the correct behavior, at least.
